### PR TITLE
#8293 Unavailable overlay for sidebar forms and temp panels on nav

### LINKED ISSUE #8284: Sidebar form doesn't submit
## Describe the bug

- https://pixiebrix.slack.com/archives/C0436P48QHY/p1713284277907749

## To Reproduce

> Not sure how to reproduce but im clicking submit in a mod and nothing is happening. I am sure if i refresh the page and load the mod again it will work.

1. install `@pixies/test/temp-form-in-sidebar`
2. open the form in the side panel
3. refresh the page
4. attempt to submit the form

## Actual behavior

- Clicking submit doesn't do anything

## Expected behavior

- Form shows an error if there's an invalid field, or
- Form submits when clicking submit

## Screenshots/Loom

- See Slack: https://pixiebrix.slack.com/archives/C0436P48QHY/p1713284277907749

## Desktop (please complete the following information):

- Browser: Chrome
- Extension Version: 1.8.13-alpha.3-mv3

## Additional context

- What mod(s) were you running?: `@pixies/internal/bug-tracker`
- What website(s) were you using?: Slack

### LINKED ISSUE #8293: Slice 2: Show “Panel is no longer available” overlay on info panels/forms
Feature UX: use the provided UX mockup https://www.figma.com/file/iK3zlhXpXqjPCX8jgriynh/SideBar?type=design&node-id=1748-3224&mode=design&t=SjRFqEO49OZWGfmF-4

Engineering Considerations/Implementation Sketch
- Add a new stale boolean attribute to TemporaryPanelEntry and FormPanelEntry, and a new reducer method invalidatePanels , for updating all current forms and panels to stale = true
- Set up listener in the [[ConnectedSidebar](https://github.com/pixiebrix/pixiebrix-extension/blob/83fd174a189408e7eb133b6b5f5d21851ef738f8/src/sidebar/ConnectedSidebar.tsx#L131-L130)](https://github.com/pixiebrix/pixiebrix-extension/blob/83fd174a189408e7eb133b6b5f5d21851ef738f8/src/sidebar/ConnectedSidebar.tsx#L131-L130) for browser.webNavigation.onBeforeNavigate (note: checking that frameType is "outermost_frame" and the tabId is the one connected to the sidebar) that sets any temporary and form panels as stale using the sidebarSlice setPanels method.
- We will need to pass the connected tabId as a prop into the sidebar component
Render the “Panel not available on this Page” overlay over the temporary and form panel body. Close button should trigger closing the panel.

Test Plan:
- Ensure that this not occur on SPA navigation (ex. on a github repo, switching from the Code to Pull Requests tab.
- If a link opens a new tab, the old tab’s sidebar panels should not be marked as stale
- Ensure that this occurs when refreshing the page
- Navigating via new link in the browser or following a link.

From: https://www.notion.so/pixiebrix/Phase-gracefully-handle-page-reload-navigation-in-the-Chromium-Side-Panel-0efdedb6c1e44b088e65106202e08c28?pvs=4#ec3a3eef571a4be988dcee641ee86759

### PR DESCRIPTION
**Pull request by fungairino**
## What does this PR do?

- Closes #8293
- Fixes #8284
- Includes a fix for the notification sound for the `npm run watch` command.
- Includes some e2e lint error fixes

## Discussion

- Perhaps we don't need a feature flag for this feature. Thoughts?

## Demo

Screenshot 2024-04-25 at 8 44 44 PM

https://github.com/pixiebrix/pixiebrix-extension/assets/3498063/184394a3-23eb-4ae3-8c73-956861483e13

## TODO

- [X] Get the close button working
- [x] Add to strict null checks
- [x] Playwright test

## Checklist

- [x] Add jest or playwright tests and/or storybook stories
- [X] Designate a primary reviewer @BLoe

### CODE REVIEW

File: src/sidebar/ConnectedSidebar.tsx [hunk 1]
```diff
@@ -98,7 +104,21 @@ const ConnectedSidebar: React.VFC = () => {
   const listener = useConnectedListener();
   const sidebarIsEmpty = useSelector(selectIsSidebarEmpty);
 
-  // `useAsyncEffect` will run once on component mount since listener and formsRef don't change on renders.
+  const navigationListener = useCallback(
```
Comment (twschiller) [inline src/sidebar/ConnectedSidebar.tsx:107]: For readability, I'd recommend moving this a separate hook because it's an independent concern. As-is, it's weaved in with the other listener registration code. E.g., `useHostNavigationListener()`

File: src/sidebar/ConnectedSidebar.tsx [hunk 2]
```diff
@@ -132,10 +152,13 @@ const ConnectedSidebar: React.VFC = () => {
     // To avoid races with panel registration, listen after reserving the initial panels.
     addListener(listener);
 
+    browser.webNavigation.onBeforeNavigate.addListener(navigationListener);
```
Comment (twschiller) [inline src/sidebar/ConnectedSidebar.tsx:155]: You'll probably need to add mocks for tests. 

Also, instead of putting the MV3 condition in the handler, I'd recommend conditionally adding the listener instead

File: src/sidebar/Tabs.module.scss [hunk 3]
```diff
@@ -24,6 +24,7 @@
 
 .tabContainer {
   flex-wrap: nowrap;
+  position: relative;
```
Comment (twschiller) [inline src/sidebar/Tabs.module.scss:27]: What's the change for? Will this break any overflow/scroll logic inside panels?

File: src/sidebar/unavailableOverlay.module.scss [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+  border: 0;
+}
+
+.modal-body {
+  background: white;
+  border: 0;
+}
+
+.modal-button {
+  border-radius: 4px;
```
Comment (twschiller) [inline src/sidebar/unavailableOverlay.module.scss:56]: @BrandonPxBx is this a different border radius than other buttons in the UI? Is that intentional?

File: src/sidebar/UnavailableOverlay.tsx [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+import React from "react";
+import styles from "./unavailableOverlay.module.scss";
+import cx from "classnames";
+import { Button, Modal } from "react-bootstrap";
+
+const UnavailableOverlay = () => (
+  <div className={styles["unavailable-overlay"]}>
+    <div
+      className={cx("modal show position-static w-auto h-auto")}
+      style={{ display: "block", marginTop: "10vh" }}
```
Comment (twschiller) [inline src/sidebar/UnavailableOverlay.tsx:27]: Why not use CSS  here?

File: src/sidebar/UnavailableOverlay.tsx [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    >
+      <Modal.Dialog
+        size={"sm"}
```
Comment (twschiller) [inline src/sidebar/UnavailableOverlay.tsx:30]: NIT: `{` not required for string constants

File: src/sidebar/UnavailableOverlay.tsx [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import React from "react";
+import styles from "./unavailableOverlay.module.scss";
+import cx from "classnames";
+import { Button, Modal } from "react-bootstrap";
+
+const UnavailableOverlay = () => (
+  <div className={styles["unavailable-overlay"]}>
```
Comment (twschiller) [inline src/sidebar/UnavailableOverlay.tsx:24]: IIRC, in other places we write using `unavailableOverlay` so we can just do `styles.unavailableOverlay`. Any reason to prefer kebab-case?

File: src/sidebar/UnavailableOverlay.tsx [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+    <div
+      className={cx("modal show position-static w-auto h-auto")}
+      style={{ display: "block", marginTop: "10vh" }}
+    >
+      <Modal.Dialog
+        size={"sm"}
+        className={cx(styles["modal-dialog"], "shadow text-center")}
```
Comment (twschiller) [inline src/sidebar/UnavailableOverlay.tsx:31]: NIT: I prefer not mixing utility classes and class names for a component because it splits the concern across 2 locations. Although we don't have a great way to reference the utility classes in the CSS 🤷 so perhaps the code re-use better

File: src/types/sidebarTypes.ts [hunk 9]
```diff
@@ -111,6 +111,12 @@ type BasePanelEntry = {
    * The panel type.
    */
   type: EntryType;
+
+  /**
+   * Determines if the panel cannot be displayed for the current tab. Used
+   * to show an overlay over the panel to indicate it is unavailable.
+   */
+  isUnavailable?: boolean;
```
Comment (twschiller) [inline src/types/sidebarTypes.ts:119]: NIT: adding `@since 1.8.14` can be a good way for developers to get a sense of the type/API evolution

File: webpack.config.mjs [hunk 10]
```diff
@@ -203,7 +203,7 @@ const createConfig = (env, options) =>
         }),
 
       // Only notifies when watching. `zsh-notify` is suggested for the `build` script
-      options.watch &&
+      !isProd(options) &&
```
Comment (twschiller) [inline webpack.config.mjs:206]: ❤️

File: src/types/sidebarTypes.ts (hunk 9 above)
Comment (twschiller) [reply src/types/sidebarTypes.ts:119]: Given that the value isn't persisted anywhere, should we just make it required to simplify checks? In general, avoiding nullness where possible is probably "a good thing™"

File: src/sidebar/unavailableOverlay.module.scss (hunk 4 above)
Comment (BrandonPxBx) [reply src/sidebar/unavailableOverlay.module.scss:56]: Most likely an artifact of the Figma components not matching production components. What's the normal radius on our buttons in the code base? I'll update the Figma components to match.

File: src/sidebar/Tabs.module.scss (hunk 3 above)
Comment (fungairino) [reply src/sidebar/Tabs.module.scss:27]: This is needed so that the blur overlay only covers the panel content area, and does not cover the whole sidebar (including the tabs and logo).
As far as I can tell, it has no other impact on UX such as overflow and scrolling. I can make a note here.

File: src/sidebar/UnavailableOverlay.tsx (hunk 7 above)
Comment (fungairino) [reply src/sidebar/UnavailableOverlay.tsx:24]: I've seen some inconsistent naming pattern in scss files between camel and dash case for class names, but I'm happy to switch back to camel case

File: src/sidebar/UnavailableOverlay.tsx (hunk 5 above)
Comment (fungairino) [reply src/sidebar/UnavailableOverlay.tsx:27]: I can move it into the scss file, just missed it

File: src/sidebar/UnavailableOverlay.tsx (hunk 8 above)
Comment (fungairino) [reply src/sidebar/UnavailableOverlay.tsx:31]: Yea, my concern is re-usability and consistency. I'll see if it looks better if I move it into the styles file.

File: src/sidebar/UnavailableOverlay.tsx (hunk 7 above)
Comment (BLoe) [reply src/sidebar/UnavailableOverlay.tsx:24]: IMO - CSS Modules should use camel case because the classes are used directly in js code, and regular scss files should prefer kebab case because that's what will actually be added to the DOM attribute.

File: src/sidebar/UnavailableOverlay.tsx (hunk 8 above)
Comment (BLoe) [reply src/sidebar/UnavailableOverlay.tsx:31]: Yeah, we do this in a lot of places already

File: src/types/sidebarTypes.ts (hunk 9 above)
Comment (BLoe) [reply src/types/sidebarTypes.ts:119]: Does it need to be designed as a negative boolean? Any chance we can go with `isAvailable` instead for simplicity's sake? Does that make the logic worse or something?

File: src/sidebar/unavailableOverlay.module.scss (hunk 4 above)
Comment (fungairino) [reply src/sidebar/unavailableOverlay.module.scss:56]: The walkthrough modal here (accessed from the quickbar, "Learn: Open the page Editor"):
Screenshot 2024-04-26 at 12 13 50 PM

Is dynamic for some reason, and is around 4px border radius

Other modal components in our storybook are at 4.8px
https://pixiebrix-extension-storybook.vercel.app/?path=/story/components-confirmationmodal--custom-content

File: src/types/sidebarTypes.ts (hunk 9 above)
Comment (fungairino) [reply src/types/sidebarTypes.ts:119]: It's easier to make this not required since hunting down all the definitions and factories and updating them to include an explicit isUnavailable attribute is somewhat tedious and slightly impacts readability. As you noted, it's almost functionally the same, except for cases when for some reason, we would want to check for the existence of the attribute itself in the type.

File: src/sidebar/unavailableOverlay.module.scss (hunk 4 above)
Comment (BrandonPxBx) [reply src/sidebar/unavailableOverlay.module.scss:56]: Looks like there's stuff all over the place in the product.

Default bootstrap settings
* Large - 0.3 rem (4.8 px @ 16px default font size)
* Small - 0.2 rem (3.2 px @ 16px default font size)

Some other buttons
* Refresh icon button on Integration Select filed in Page Editor - 4px hardcoded
* Refresh button on the Preview Pane in the Data Panel on a Trigger Starter Brick - 3.2px hard coded

Because the "system" in Figma is based on the Bootstrap theme, but we cannot use REM values as measurements in Figma, I think there's been some lost in translation throughout the design life cycle.

TL;DR - Lets go with default bootstrap of 0.2 rem on the small buttons.

Comment (fungairino) [reply src/sidebar/unavailableOverlay.module.scss:56]: What about the border-radius for the modal dialog container? In the mock-up you used 12px, but that doesn't seem to line up with the other modals and bootstrap default.

File: src/sidebar/Tabs.module.scss (hunk 3 above)
Comment (BLoe) [reply src/sidebar/Tabs.module.scss:27]: I'm far from a CSS expert, but it makes sense to me that a "container" element would have `position: relative`, right? As in this exact use-case, elements inside the "container" should expect to be able to declare there position relative to the container rather than the entire sidebar window.

File: src/types/sidebarTypes.ts (hunk 9 above)
Comment (BLoe) [reply src/types/sidebarTypes.ts:119]: That makes sense to me as long as it doesn't make strict-null-checks substantially harder in some way

File: src/store/sidebar/thunks/removeFormPanel.ts [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+    removedEntry,
+    forms: otherFormPanels,
+  };
+});
+
+/**
+ * Close form panels
+ * @param nonces panel nonces
+ */
+async function closeForms(nonces: UUID[]): Promise<void> {
```
Comment (twschiller) [inline src/store/sidebar/thunks/removeFormPanel.ts:60]: I don't think calling cancelForm is necessary given that the forms won't exist on the form controller after the page redirect/navigation (because there will be a new content script)

File: end-to-end-tests/tests/bricks/sidebarEffects.spec.ts [hunk 12]
```diff
@@ -32,7 +32,6 @@ test.describe("sidebar effect bricks", () => {
     await page.goto("/");
 
     // Ensure the page is focused by clicking on an element before running the keyboard shortcut, see runModViaQuickbar
-    await page.getByText("Index of  /").click();
```
Comment (fungairino) [inline end-to-end-tests/tests/bricks/sidebarEffects.spec.ts:35]: I refactored the runModViaQuickBar method so we don't have to have these `click` calls before it.

File: end-to-end-tests/tests/runtime/sidebarController.spec.ts [hunk 13]
```diff
@@ -70,7 +70,7 @@ test.describe("sidebar controller", () => {
 
     // The focus dialog should not be shown in the iframe. Check after checking the top-level frame
     // because it's a positive check for the dialog being shown.
-    await expect(frame.getByRole("button", { name: "OK" })).not.toBeVisible();
+    await expect(frame.getByRole("button", { name: "OK" })).toBeHidden();
```
Comment (fungairino) [inline end-to-end-tests/tests/runtime/sidebarController.spec.ts:73]: Unrelated lint error fixed

File: end-to-end-tests/tests/telemetry/errors.spec.ts [hunk 14]
```diff
@@ -1,10 +1,37 @@
...
-import { type Page, test as base } from "@playwright/test";
+import { type BrowserContext, type Page, test as base } from "@playwright/test";
 import { getBaseExtensionConsoleUrl } from "../../pageObjects/constants";
 import { MV } from "../../env";
 
+async function waitForBackgroundPageRequest(
+  context: BrowserContext,
+  extensionId: string,
+  errorServiceEndpoint: string,
+) {
```
Comment (fungairino) [inline end-to-end-tests/tests/telemetry/errors.spec.ts:11]: refactored this method in the process of fixing an eslint error for no conditionals

File: end-to-end-tests/utils.ts [hunk 15]
```diff
@@ -161,6 +164,15 @@ export async function waitForSelectionMenuReadiness(page: Page) {
   }).toPass({ timeout: 5000 });
 }
 
+// Waits for the quick bar to be ready to use
+async function waitForQuickBarReadiness(page: Page) {
```
Comment (fungairino) [inline end-to-end-tests/utils.ts:168]: Without this check, sometimes the quickbar would not open when immediately attempting to open it after loading a new page.

File: src/store/sidebar/thunks/removeFormPanel.ts (hunk 11 above)
Comment (fungairino) [reply src/store/sidebar/thunks/removeFormPanel.ts:60]: In that case, it will be a no-op. See: https://github.com/pixiebrix/pixiebrix-extension/blob/bdab1e60581ab042fa6f9cf65e4cd823b81c0cbe/src/platform/forms/formController.ts#L132

I guess we could check the panel and if `panel.isUnavailable` then we skip calling closeForms

File: src/sidebar/ConnectedSidebar.test.tsx [hunk 16]
```diff
@@ -98,6 +107,45 @@ describe("SidebarApp", () => {
...
+          },
+        },
+      );
+
+      await waitForEffect();
+
+      // The navigation listener should be added for MV3
+      expect(
+        browser.webNavigation.onBeforeNavigate.addListener,
+      ).toHaveBeenCalledWith(expect.any(Function));
```
Comment (BLoe) [inline src/sidebar/ConnectedSidebar.test.tsx:141]: Does this have to use `waitForEffect()`? Could this be done instead as:
```suggestion
      await waitFor(() => {
         // The navigation listener should be added for MV3
        expect(
          browser.webNavigation.onBeforeNavigate.addListener,
        ).toHaveBeenCalledWith(expect.any(Function));
      });
```

File: src/store/sidebar/thunks/removeFormPanel.ts (hunk 11 above)
Comment (BLoe) [reply src/store/sidebar/thunks/removeFormPanel.ts:64]: The form/panel registration logic already needs to handle invalid/unregistered nonces in order to pass strict null checks, so we can be confident that a bug will not be introduced here in the future. Adding the if-check here means that this thunk is now essentially tightly coupled with behavior in other parts of the redux slice, with no change in functionality of this thunk. I think we should remove the if-check here and let `cancelForm` no-op, since it's simpler and less-coupled overall code, with no increased risk.


@twschiller This is also used when the user clicks the close button on the panel, not just on navigation/with unavailable panels.

File: src/store/sidebar/thunks/removeFormPanel.ts [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+  const topLevelFrame = await getConnectedTarget();
+  cancelForm(topLevelFrame, ...nonces);
```
Comment (BLoe) [inline src/store/sidebar/thunks/removeFormPanel.ts:66]: Any particular reason to extract these two lines into a separate, local (private) function here? Why not just inline this?

File: src/testUtils/factories/sidebarEntryFactories.ts [hunk 18]
```diff
@@ -38,11 +38,13 @@ const activateModPanelEntryFactory = define<ModActivationPanelEntry>({
     },
   ],
   heading: (n: number) => `Activate Mods Test ${n}`,
+  isUnavailable: false,
```
Comment (BLoe) [inline src/testUtils/factories/sidebarEntryFactories.ts:41]: Just curious, if this isn't required, then we can probably leave these values off the factories, right?

**Review by BLoe — COMMENTED** [review]
I'm not the one who should approve the design/UX parts here, but this is looking good from a code perspective to me 👍 Happy to unblock if needed but will defer to someone else here to approve based on the design/copy.

File: src/testUtils/factories/sidebarEntryFactories.ts (hunk 18 above)
Comment (fungairino) [reply src/testUtils/factories/sidebarEntryFactories.ts:41]: Yes, we could.

File: src/store/sidebar/thunks/removeFormPanel.ts (hunk 17 above)
Comment (fungairino) [reply src/store/sidebar/thunks/removeFormPanel.ts:66]: Mostly just following the pattern set by the other `removeTemporaryPanel.ts` thunk, but I agree it can be inlined

File: src/sidebar/ConnectedSidebar.test.tsx (hunk 16 above)
Comment (fungairino) [reply src/sidebar/ConnectedSidebar.test.tsx:141]: Turns out it's not really even needed

**Comment by fungairino** [comment]
> I'm not the one who should approve the design/UX parts here
@BrandonPxBx has reviewed the final UX!
